synthesizer/src/vm/finalize.rs

```diff
@@ -28,31 +30,75 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     ///     `Ratify::BlockReward(block_reward)` and `Ratify::PuzzleReward(puzzle_reward)`
     ///     to the front of the `ratifications` list.
     #[inline]
-    pub fn speculate<'a>(
+    pub fn speculate<'a, R: Rng + CryptoRng>(
         &self,
         state: FinalizeGlobalState,
         coinbase_reward: Option<u64>,
         candidate_ratifications: Vec<Ratify<N>>,
         candidate_solutions: &Solutions<N>,
         candidate_transactions: impl ExactSizeIterator<Item = &'a Transaction<N>>,
+        rng: &mut R,
     ) -> Result<(Ratifications<N>, Transactions<N>, Vec<N::TransactionID>, Vec<FinalizeOperation<N>>)> {
         let timer = timer!("VM::speculate");
 
+        // TODO (raychu86): Clean up this logic and remove the extra collect.
+        // If the transactions are not part of the genesis block, ensure each transaction is well-formed and unique. Abort any transactions that are not.
+        let block_store = self.block_store();
+        let candidate_transactions = candidate_transactions.collect::<Vec<_>>();
+        let (verified_transactions, verification_aborted_transactions) = match block_store
+            .find_block_height_from_state_root(block_store.current_state_root())?
+        {
+            // If the current state root does not exist in the block store, then the genesis block has not been introduced yet.
+            None => (candidate_transactions, vec![]),
+            // Verify transactions for all non-genesis cases.
+            _ => {
+                let rngs = (0..candidate_transactions.len()).map(|_| StdRng::from_seed(rng.gen())).collect::<Vec<_>>();
+                // Verify the transactions and collect the error message if there is one.
+                let checked_transactions: Vec<_> = cfg_into_iter!(candidate_transactions)
```

Comment:
This comment still applies: https://github.com/AleoHQ/snarkOS/pull/3083#issuecomment-1936937113
I was able to kill all the nodes with some big batch proposals in a test suite. Recall:
- The maximum RAM to verify a 1M constraint program is 2GB.
- The maximum RAM for a single call execution is 17MB
- The maximum RAM for a 31-depth call execution is <currently looking into it>

A thorough but brittle approach would be to estimate memory usage based on the number of constraints. But an easier setup would be a separate deploy and an execute verification maximum:

```
const MAX_PARALLEL_DEPLOY_VERIFICATIONS = 10;
const MAX_PARALLEL_EXECUTE_VERIFICATIONS = <currently looking into it>;

let (deploys, executions) = candidate_transactions.partition(|tx|tx.is_deploy());
let deploy_iter = deploys.chunks(MAX_PARALLEL_DEPLOY_VERIFICATIONS);
let execution_iter = executions.chunks(MAX_PARALLEL_EXECUTE_VERIFICATIONS);

while let Some(deploys) = deploy_iter.next() {
    cfg_into_iter!(deploys).zip(rngs).map(|| ...verify...)
}
while let Some(executions) = execution_iter.next() {
    cfg_into_iter!(executions).zip(rngs).map(|| ...verify...)
}
```

Reply:
How is this different from the current approach of transaction verification on the current `mainnet` branch in `check_next_block`?

Are you saying that fundamentally the verification logic should be updated?

Reply:
Yes, that piece of parallel verification logic was prone to the same vulnerability.

Reply:
I've implemented the OOM fix here - https://github.com/AleoHQ/snarkVM/pull/2387. If it is approved, we can apply the change to this PR as well.